plugins/rest/rest_auth.go

```diff
@@ -113,56 +144,148 @@ type tokenEndpointResponse struct {
 // oauth2ClientCredentialsAuthPlugin represents authentication via a bearer token in the HTTP Authorization header
 // obtained through the OAuth2 client credentials flow
 type oauth2ClientCredentialsAuthPlugin struct {
-	TokenURL     string    `json:"token_url"`
-	ClientID     string    `json:"client_id"`
-	ClientSecret string    `json:"client_secret"`
-	Scopes       *[]string `json:"scopes,omitempty"`
-
-	tokenCache    *oauth2Token
-	tlsSkipVerify bool
+	GrantType    string                 `json:"grant_type"`
+	TokenURL     string                 `json:"token_url"`
+	ClientID     string                 `json:"client_id"`
+	ClientSecret string                 `json:"client_secret"`
+	SigningKeyID string                 `json:"signing_key"`
+	Claims       map[string]interface{} `json:"additional_claims"`
+	IncludeJti   bool                   `json:"include_jti_claim"`
+	Scopes       []string               `json:"scopes,omitempty"`
+
+	signingKey       *keys.Config
+	signingKeyParsed interface{}
+	tokenCache       *oauth2Token
+	tlsSkipVerify    bool
 }
 
 type oauth2Token struct {
 	Token     string
 	ExpiresAt time.Time
 }
 
+func (ap *oauth2ClientCredentialsAuthPlugin) createAuthJWT(claims map[string]interface{}, signingKey interface{}) (*string, error) {
+	now := time.Now()
+	baseClaims := map[string]interface{}{
+		"iat": now.Unix(),
+		"exp": now.Add(10 * time.Minute).Unix(),
+	}
+	for k, v := range baseClaims {
+		claims[k] = v
+	}
+
+	if len(ap.Scopes) > 0 {
+		claims["scope"] = strings.Join(ap.Scopes, " ")
+	}
+
+	if ap.IncludeJti {
+		jti, err := uuid.New(rand.Reader)
+		if err != nil {
+			return nil, err
+		}
+		claims["jti"] = jti
+	}
+
+	headers := fmt.Sprintf(`{"typ":"JWT","alg":"%v"}`, ap.signingKey.Algorithm)
+	payload, err := json.Marshal(claims)
+	if err != nil {
+		return nil, err
+	}
+
+	standardHeaders := &jws.StandardHeaders{}
+	jwsHeaders := []byte(headers)
```

Comment:
Minor details, but I guess this (not row 190) could be replace by:

```
	standardHeaders := &jws.StandardHeaders{Type: "JWT"}
	standardHeaders.Set("alg", ap.signingKey.Algorithm)
	jwsHeaders, err := json.Marshal(standardHeaders)
	if err != nil {
		return nil, err
	}
```

Reply:
Latest push eliminates the need for that struct entirely.